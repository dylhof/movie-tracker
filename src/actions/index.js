export const storeMovies = (movies) => ({
  type: 'STORE_MOVIES',
  movies
})

export const setCurrentUser = (name, id) => ({
  type: 'SET_USER',
  name,
  id
})

export const logoutCurrentUser = () => ({
  type: 'LOGOUT_USER'
})

export const addFavorite = (id) => ({
  type: 'ADD_FAVORITE',
  id
})

export const addAllUserFavorites = (favorites) => ({
  type: 'ADD_ALL_USER_FAVORITES',
  favorites
})

export const deleteFavorite = (id) => ({
  type: 'DELETE_FAVORITE',
  id
})

export const clearFavorites = () => ({
  type: 'CLEAR_FAVORITES',
})

export const setLoading = (bool) => ({
  type: 'SET_LOADING',
  bool
})