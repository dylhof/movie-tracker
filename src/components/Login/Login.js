import React, { Component } from 'react';

export default class Login extends Component {
  constructor() {
    super()
    this.state = {
      username: '',
      password: '',
    }
  }
  handleChange = (event) => {
    if (event.target.name === 'username') {
      this.setState({ username: event.target.value })
    } else {
      this.setState({ password: event.target.value })
    }
  }

  //   sign in   /api/users
  // Create Account - /api/users / new
  //   Add Favorite - /api/users / favorites / new
  //   Receive All Favorites - /api/users /: user_id / favorites
  // Delete a Favorite - /api/users /: user_id / favorites /: movie_id

  handleSubmit = async (event) => {
    event.preventDefault();
    const url = 'http://localhost:3000/api/users/new'
    try {
      const response = await fetch(url, {
        method: 'POST',
        body: JSON.stringify({ name: 'Matt', email: this.state.username, password: this.state.password }),
        headers: {
          'Content-Type': 'application/json'
        }
      })
      const result = await response.json()
    }
    catch (error) {
      throw Error(`There was an error: ${Error.status}`)
    }
  }

  render() {
    return (
      <form onSubmit={this.handleSubmit}>
        <input name='username' value={this.state.username} onChange={this.handleChange} />
        <input name='password' value={this.state.password} onChange={this.handleChange} />
        <button>Submit</button>
      </form>
    )
  }


}

// login form:
// local state to store what they are typing in (control form)
// username field
// password field
// submit/ login button
// link to sign up form


// sign up form:
// local storage (control form)
// name field
// username field
// pw field
// submit/signup
// link to log in