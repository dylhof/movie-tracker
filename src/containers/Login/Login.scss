.login-form-div {
width: 100%;
display: flex;
justify-content: center;
margin-top: 20px;

.login-form {
    display: flex;
    border-radius: 10px;
    box-shadow: 5px 5px 7px #2b0c0e;  
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    background: #0D0707;
    height: 300px;
    width: 50%;
  }
}

.login-inner-div {
  padding: 3px;
  display: flex;
  flex-direction: column;
  width: 60%;
}
.login-input {
  height: 40px;
  font-size: 25px;
  background-color: #f8e5d2;
  width: 100%;
  padding: 0px;
  border: 0px;
  border-radius: 3px; 
  padding-left: 5px;
  &:focus{
    background-color: $font-color;
    outline: 0px;
  }
}

.login-label {
  color: $font-color;
  margin-bottom: 3px;
}

.login-submit {
  width: 60%;
  height: 40px;
  padding: 0px;
  border: 0px;
  margin-top: 20px;
  background-color: #f8e5d2; 
  border-radius: 3px; 
  font-size: 20px;

  &:hover{
    background-color: $font-color;
  }
  &:focus{
    outline: 0px;
  }
}