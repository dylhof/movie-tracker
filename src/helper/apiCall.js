export const fetchData = async (url) => {
  const response = await fetch(url)
  if(response.ok) {
    return response.json() 
  } else {
    throw new Error(`Error fetching, code: ${response.status}`)
  }
}

export const fetchPost = async (url, options) => {
  const response = await fetch(url, options)
  if (response.ok) {
    return response.json()
  } else {
    throw new Error(response.status)
  }
}