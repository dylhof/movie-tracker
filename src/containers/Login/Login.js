import React, { Component } from 'react';
import { fetchPost, fetchData } from '../../helper/apiCall'
import { setCurrentUser, addAllUserFavorites } from '../../actions';
import { connect } from 'react-redux';
import PropTypes from 'prop-types'

export class Login extends Component {
  constructor() {
    super()
    this.state = {
      username: '',
      password: ''
    }
  }
  handleChange = (event) => {
    if (event.target.name === 'username') {
      this.setState({ username: event.target.value })
    } else {
      this.setState({ password: event.target.value })
    }
  }

  handleSubmit = async (event) => {
    event.preventDefault();
    try {
      const response = await fetchPost('http://localhost:3000/api/users',
        {
          method: 'POST',
          body: JSON.stringify({ email: this.state.username.toLowerCase(), password: this.state.password }),
          headers: {
            'Content-Type': 'application/json'
          }
        })
      const user_id = response.data.id
      this.props.dispatchSetCurrentUser(response.data.name, user_id)
      const url = `http://localhost:3000/api/users/${user_id}/favorites`
      const favoritesResponse = await fetchData(url)
      const favoriteIDs = favoritesResponse.data.map(favorite => favorite.movie_id)
      this.props.dispatchAddAllUserFavorites(favoriteIDs)
    } catch (error) {
      if (error.message === '500') {
        this.setState({ error: "Email or Password doesn't match" })
      }
    }
  }

  render() {
    return (
      <div className='login-form-div'>
      <form className='login-form' onSubmit={this.handleSubmit}>
        <div className='login-inner-div'>
          <label for='login-email' className='login-label'>Email</label>
          <input id='login-email' className='login-input' name='username' value={this.state.username} onChange={this.handleChange} type='email'/>
        </div>
        <div className='login-inner-div'>
          <label for='login-password' className='login-label'>Password</label>
          <input id='login-password' className='login-input' name='password' value={this.state.password} onChange={this.handleChange} />
        </div>
        <button className='login-submit'>Submit</button>
        <span>{this.state.error}</span>
      </form>
      </div>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  dispatchSetCurrentUser: (name, id) => dispatch(setCurrentUser(name, id)),
  dispatchAddAllUserFavorites: (favorites) => dispatch(addAllUserFavorites(favorites))
})

export default connect(null, mapDispatchToProps)(Login)

Login.propTypes = {
  dispatchSetCurrentUser: PropTypes.func,
  dispatchAddAllUserFavorites: PropTypes.func
}