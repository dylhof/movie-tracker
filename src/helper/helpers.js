import * as API from './apiCall';

export const tryUnfavorite = async (id, userID) => {

  const url = `http://localhost:3000/api/users/${userID}/favorites/${id}`
  await API.fetchPost(url,
    {
      method: 'DELETE',
      body: JSON.stringify({
        user_id: userID, movie_id: id
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

export const tryFavorite = async (title, id, userID, poster_path, release_date, vote_average, overview) => {

  await API.fetchPost('http://localhost:3000/api/users/favorites/new',
    {
      method: 'POST',
      body: JSON.stringify({
        movie_id: id, user_id: userID, title, poster_path,
        release_date, vote_average, overview
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

