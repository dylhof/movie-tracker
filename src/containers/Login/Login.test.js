import React from 'react'
import { shallow, mount } from 'enzyme'
import { Login, mapDispatchToProps } from './Login'
import * as API from '../../helper/apiCall'
import * as actions from '../../actions/index'

describe('Login', () => {
  let wrapper
  beforeEach(() => {
    const mockDispatchAddAllUserFavorites = jest.fn()
    const mockDispatchSetCurrentUser = jest.fn()
    const mockDispatchSetError = jest.fn()
    wrapper = shallow(
      <Login
        dispatchSetCurrentUser={mockDispatchSetCurrentUser}
        dispatchAddAllUserFavorites={mockDispatchAddAllUserFavorites}
        dispatchSetError={mockDispatchSetError}
      />
    )
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should have initial state', () => {
    //setup
    const expected = { username: '', password: '', error: '' }
    //expectation
    expect(wrapper.state()).toEqual(expected)
  })

  it('should call handleSubmit when submit button is clicked', () => {
    //setup
    wrapper = mount(<Login />)
    const spy = spyOn(wrapper.instance(), 'handleSubmit')
    const mockEvent = {preventDefault: jest.fn}
    wrapper.instance().forceUpdate()
    //execution
    wrapper.find('form').simulate('submit', mockEvent)
    //expectation
    expect(spy).toHaveBeenCalled()
  })

  describe('handleChange', () => {
    it('should set state of username', () => {
      //setup
      const mockEvent = { target: { name: 'username', value: 'Dylan' } }
      //execution
      wrapper.instance().handleChange(mockEvent)
      //expectation
      expect(wrapper.state('username')).toEqual('Dylan')
    })

    it('should set state of password', () => {
      //setup
      const mockEvent = { target: { name: 'password', value: 'Dylan' } }
      //execution
      wrapper.instance().handleChange(mockEvent)
      //expectation
      expect(wrapper.state('password')).toEqual('Dylan')
    })
  })

  describe('handleSubmit', () => {
    let mockEvent
    beforeEach(() => {
      mockEvent = Object.assign(jest.fn(), { preventDefault: () => { } })
    })
    it('should call fetchPost', () => {
      //setup
      API.fetchPost = jest.fn()
      //execution
      wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(API.fetchPost).toHaveBeenCalled()
    })

    it('should call dispatchSetCurrentUser with the correct params', async () => {
      //setup
      const mockResponse = { data: { name: 'Matt', id: 2 } }
      API.fetchPost = jest.fn().mockImplementation(() => Promise.resolve(mockResponse))
      //execution
      await wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchSetCurrentUser).toHaveBeenCalledWith('Matt', 2)
    })

    it('should call fetchData', async () => {
      //setup
      API.fetchData = jest.fn()
      const mockResponse = { data: { name: 'Matt', id: 2 } }
      API.fetchPost = jest.fn().mockImplementation(() => Promise.resolve(mockResponse))
      //execution
      await wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(API.fetchData).toHaveBeenCalledWith('http://localhost:3000/api/users/2/favorites')
    })

    it('should call dispatchAddAllUserFavorites with the correct params', async () => {
      //setup
      const mockResponse = { data: { name: 'Matt', id: 2 } }
      const mockFavoritesData = { data: [{ title: 'AquaMan', movie_id: 1 }, { title: 'Serenity', movie_id: 3 }] }
      API.fetchPost = jest.fn().mockImplementation(() => Promise.resolve(mockResponse))
      API.fetchData = jest.fn().mockImplementation(() => Promise.resolve(mockFavoritesData))
      //execution
      await wrapper.instance().handleSubmit(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchAddAllUserFavorites).toHaveBeenCalledWith([1, 3])
    })
  })

  describe('mapDispatchToProps', () => {

    it('should call dispatch when calling dispatchSetCurrentUser from mdtp', () => {
      //setup
      const mockDispatch = jest.fn()
      const actionToDispatch = actions.setCurrentUser('Matt', 1)
      //execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.dispatchSetCurrentUser('Matt', 1)
      //expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch when calling dispatchAddAllUserFavorites from mdtp', () => {
      //setup
      const mockDispatch = jest.fn()
      const actionToDispatch = actions.addAllUserFavorites([1, 3])
      //execution
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.dispatchAddAllUserFavorites([1, 3])
      //expectation
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})