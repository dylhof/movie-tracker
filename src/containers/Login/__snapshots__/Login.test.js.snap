// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<div
  className="login-form-div"
>
  <form
    className="login-form"
    onSubmit={[Function]}
  >
    <div
      className="login-inner-div"
    />
    <div
      className="login-inner-div"
    >
      <label
        className="login-label"
        htmlFor="login-email"
      >
        Email
      </label>
      <input
        className="login-input"
        id="login-email"
        name="username"
        onChange={[Function]}
        type="email"
        value=""
      />
    </div>
    <div
      className="login-inner-div"
    >
      <label
        className="login-label"
        htmlFor="login-password"
      >
        Password
      </label>
      <input
        className="login-input"
        id="login-password"
        name="password"
        onChange={[Function]}
        value=""
      />
    </div>
    <button
      className="login-submit"
    >
      Submit
    </button>
    <p
      className="signup-message"
    >
      Don't have an account? 
      <Link
        replace={false}
        to="/signup"
      >
        Sign up!
      </Link>
    </p>
  </form>
</div>
`;
